### PR TITLE
Modernise packaging

### PR DESCRIPTION
**Pull request by ZedThree**
- Modernises the packaging, using `pyproject.toml` for config
- Sets the version automatically from the git tag using `setuptools_scm`
- Applies black formatting and some ruff fixes
- Adds automated workflows for testing, linting, and publishing

The automated formatting is currently just a check, but it is also possible to automatically apply the formatting too.

To finish setting up the publishing, see the guide here: https://docs.pypi.org/trusted-publishers/creating-a-project-through-oidc/
Basically, you just need to create a project on PyPI in a few steps.

### CODE REVIEW

File: src/simplesimdb/__init__.py [hunk 1]
```diff
@@ -554,38 +585,37 @@ def delete( self, js, n = 0) :
...
-        for entry in files :
-            if entry["n"] == 0 :
-                os.remove( entry["inputfile"])
-            os.remove( entry["outputfile"])
-        registry = dict()
-        self.set_registry( registry)
-        try :
-            os.rmdir( self.__directory)
-        except OSError as e:
-            pass # if the directory is non-empty nothing happens
```
Comment (mwiesenberger) [inline src/simplesimdb/__init__.py]: Please keep comment

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  "pytest",
+]
+lint = [
+  "ruff",
+]
+
+[build-system]
+requires = [
+    "setuptools >= 65",
+    "setuptools_scm >= 8.0.0",
```
Comment (mwiesenberger) [inline pyproject.toml]: Here,
https://setuptools-scm.readthedocs.io/en/latest/usage/#semantic-versioning-semver
it writes "setuptools-scm" instead of underscore "setuptools_scm"
Not sure if this is equivalent?

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+[project.urls]
+Source = "https://github.com/mwiesenberger/simplesimdb"
+Tracker = "https://github.com/mwiesenberger/simplesimdb/issues"
+
+[project.optional-dependencies]
+test = [
+  "pytest",
+]
+lint = [
```
Comment (mwiesenberger) [inline pyproject.toml]: Is this entry necessary?
i.e. is `pip install -e .[lint]`
something we would like call?

**Comment by mwiesenberger** [comment]
I also thought a bit about moving the source file into src/simplesimdb and the test file into the tests/ folder and I think I prefer the original layout as a python module where everything sits at the base directory (since there is only a single source and test file) It like the simplicity in it

**Comment by LiamPattinson** [comment]
Hi, sorry for the lack of progress here! I'm doing a little work now to address your comments, and I'll hopefully have this PR cleaned up shortly.

File: src/simplesimdb/__init__.py (hunk 1 above)
Comment (LiamPattinson) [reply src/simplesimdb/__init__.py]: I've put it back in [1a3e094](https://github.com/mwiesenberger/simplesimdb/pull/3/commits/1a3e094ec80194a96aaedd4228883a9e97d400ee)

File: pyproject.toml (hunk 2 above)
Comment (LiamPattinson) [reply pyproject.toml]: This is something I've always found a bit confusing. Python sometimes allows you to be flexible with the choice of underscores vs hyphens, but according to the people who write the Python standards the 'canonical' project name should always be the version with hyphens, as you've indicated ([PEP 503](https://peps.python.org/pep-0503/#normalized-names)). I've updated it in [4ffc882](https://github.com/mwiesenberger/simplesimdb/pull/3/commits/4ffc882a74ab07740ef2225af442033450cb4339) to be consistent.

File: pyproject.toml (hunk 3 above)
Comment (LiamPattinson) [reply pyproject.toml]: It's something I often do in my own work, and it gives other developers a nice shorthand way to install the project and grab all of their development tools in one go.

However, when it comes to tools such as linters and formatters, I've been coming around to the idea that the correct approach isn't for developers to install them on a project-by-project basis, but instead to install/run them in isolated environments using [`pipx`](https://pipx.pypa.io/latest/) or (preferably) [`uvx`](https://docs.astral.sh/uv/guides/tools/). I've therefore removed this group in [9ff0483](https://github.com/mwiesenberger/simplesimdb/pull/3/commits/9ff0483f711b331f3b4a9cf437b676b29164eb1c).

I would argue in favour of keeping the `test` group however, as testing dependencies can vary significantly from one project to the next (with the exception of `pytest` of course).

**Comment by LiamPattinson** [comment]
> I also thought a bit about moving the source file into src/simplesimdb and the test file into the tests/ folder and I think I prefer the original layout as a python module where everything sits at the base directory (since there is only a single source and test file) It like the simplicity in it

After some experiments, I don't think it's possible to achieve this with modern setuptools using `pyproject.toml`. You can specify individual Python modules with the `py-modules` setting, similarly to how you could set `py_modules` in `setup.py`, but I wasn't able to get that to work. Additionally, `setuptools-scm` was writing a new file `src/simplesimdb/_version.py` at the point of installation, and I wasn't able to get that to work cleanly with a single-module package.

I checked around some other package managers, and it turns out [Hatch](https://hatch.pypa.io/latest/build/) gives users the option to make single-module packages. [`hatch-vcs`](https://github.com/ofek/hatch-vcs) can take the place of `setuptools-scm` and it doesn't need to write a separate version file, so this setup seems to do the trick:

```toml
[build-system]
requires = [
    "hatchling",
    "hatch-vcs",
]
build-backend = "hatchling.build"

[tool.hatch.build.targets.sdist]
include = ["simplesimdb.py"]

[tool.hatch.version]
source = "vcs"
```

This should work even if you don't use Hatch as a Python environment manager, as pretty much all build backends should be interchangeable nowadays.

File: pyproject.toml (hunk 2 above)
Comment (mwiesenberger) [reply pyproject.toml]: perfect

File: pyproject.toml (hunk 3 above)
Comment (mwiesenberger) [reply pyproject.toml]: ok perfect

**Comment by mwiesenberger** [comment]
Cool thanks for the work guys I appreciate it
